packages/modules/internal_openwb/chargepoint_module.py

```diff
@@ -0,0 +1,168 @@
+
+    def get_pins_phase_switch(self, new_phases: int) -> Tuple[int, int]:
+        # return gpio_cp, gpio_relay
+        if self.duo_num == 1:
+            return 22, 29 if new_phases == 1 else 37
+        else:
+            return 15, 11 if new_phases == 1 else 13
+
+    def get_pins_cp_interruption(self) -> int:
+        # return gpio_cp, gpio_relay
+        if self.duo_num == 1:
+            return 22
+        else:
+            return 15
+
+
+class ChargepointModule(AbstractChargepoint):
+    def __init__(self, config: InternalOpenWB) -> None:
+        self.config = config
+        self.component_info = ComponentInfo(
+            self.config.id,
+            "Ladepunkt "+str(self.config.id), "chargepoint")
+        self.__store = get_chargepoint_value_store(self.config.id)
+        self.__client = ClientFactory(self.config.id, self.config.serial_client)
+
+    def set_current(self, current: float) -> None:
+        with SingleComponentUpdateContext(self.component_info):
+            if self.set_current_evse != current:
+                self.__client.evse_client.set_current(int(current))
+
+    def get_values(self) -> Tuple[ChargepointState, float]:
+        try:
+            _, power = self.__client.meter_client.get_power()
+            if power < 10:
+                power = 0
+            voltages = self.__client.meter_client.get_voltages()
+            currents = self.__client.meter_client.get_currents()
+            imported = self.__client.meter_client.get_imported()
+            phases_in_use = sum(1 for current in currents if current > 3)
+
+            time.sleep(0.1)
+            plug_state, charge_state, self.set_current_evse = self.__client.evse_client.get_plug_charge_state()
+            self.__client.read_error = 0
+
+            rfid = ramdisk_read("readtag")
+            # reset tag
+            if rfid != "0" and plug_state is False:
+                ramdisk_write("readtag", "0")
+
+            chargepoint_state = ChargepointState(
+                power=power,
+                currents=currents,
+                imported=imported,
+                exported=0,
+                # powers=powers,
+                voltages=voltages,
+                # frequency=frequency,
+                plug_state=plug_state,
+                charge_state=charge_state,
+                phases_in_use=phases_in_use,
+                rfid=rfid
+            )
+        except Exception as e:
+            self.__client.read_error += 1
+            if self.__client.read_error > 5:
+                log.exception(
+                    "Anhaltender Fehler beim Auslesen der EVSE. Lade- und Steckerstatus werden zurückgesetzt.")
+                plug_state = False
+                charge_state = False
+                chargepoint_state = ChargepointState(
+                    plug_state=plug_state,
+                    charge_state=charge_state,
+                    phases_in_use=0
+                )
+            else:
+                raise e
```

Comment:
Dann hat man auch wieder das Problem, dass `__name__='fault_state'` ist.

Anderer Ansatz: Muss an dieser Stelle die Exception gefangen werden? 
Es gibt nicht so viele Fehlermöglichkeiten. Die Modbus-Exceptions werden im Modbus-Modul gefangen und alle anderen Module rufen die Zähler mit Single- bzw MultiComponentUpdateContext auf, dort werden die Fehler auch entsprechend behandelt.
Nur `get_values` im chargepoint_module ruft die Zählermethoden ohne SingleComponentUpdateContext auf, da der Fehler nach oben weitergereicht werden muss.
```suggestion
                raise e if isinstance (e, FaultState) else FaultState.from_exception(e)
```

Reply:
> Dann hat man auch wieder das Problem, dass `__name__='fault_state'` ist.

Da kann man einen Parameter draus machen:
```python
def exceptions_to_fault_state(module_name: str):
    def decorate(delegate: Callable):
        def wrapper(*args, **kwargs):
            try:
                return delegate(*args, **kwargs)
            except Exception as e:
                if isinstance(e, FaultState):
                    raise
                else:
                    raise FaultState.error(module_name + " " + str(type(e)) + " " + str(e)) from e
        return wrapper
    return decorate


@exceptions_to_fault_state(__name__)
def function_to_decorate():
    print("Here some stuff happens")
```

> Anderer Ansatz: Muss an dieser Stelle die Exception gefangen werden? Es gibt nicht so viele Fehlermöglichkeiten. Die Modbus-Exceptions werden im Modbus-Modul gefangen und alle anderen Module rufen die Zähler mit Single- bzw MultiComponentUpdateContext auf, dort werden die Fehler auch entsprechend behandelt. Nur `get_values` im chargepoint_module ruft die Zählermethoden ohne SingleComponentUpdateContext auf, da der Fehler nach oben weitergereicht werden muss.

Jo stimmt. Kann vielleicht auch ersatzlos gestrichen. Das ist natürlich noch einfacher.